feat: add section about Filter-based feature selection (#398)
mlr-org/mlr3book: book/optimization.qmd

* [automating](#autotuner) tuning.

We will use the `r mlr_pkg("mlr3tuning")` package, which supports common tuning operations.

> `

Reviewer: ?

filter$calculate(task)
head(as.data.table(filter), 3)
```
### Filter-based Feature Selection

A common usecase for filters is to conduct feature selection based on the filter scores.
This can be achived using `r ref("mlr_pipeops_filter", text = "PipeOpFilter")`.
This PipeOp takes as input a Task, applies the filter, and selects the features based on the calculated scores.
The method how the features are subset can be defined in four different ways, each corresponding to a different parameter:

  * `filter.nfeat` - The *nfeat* values with the highest filter score are selected.
  * `filter.frac` - The *fraction* of values with the highest filter score are selected.
  * `filter.cutoff` - All features with a filter score greater than the specified *cutoff* are selected.
  * `filter.permuted` - A random permutation of each feature is added to the task before applying the filter.
    All features selected before the *permuted*-th permuted features is selected are kept.

:::{.callout-note}
The `r ref("mlr_pipeops_filter", text = "PipeOpFilter")` only operates on the subset of features that have the correct column type and will leave other columns as is.
:::

> We will illustrate the usage of this filter with the use of a machine learning pipeline for the `r ref("mlr_tasks_optdigits", text = "optdigits")` task, where the goal is to classify handwritten digits.

Reviewer 2: ```suggestion
We will illustrate the usage of this filter on a pipeline for the `r ref("mlr_tasks_optdigits", text = "optdigits")` task, where the goal is to classify handwritten digits.
```

We will first subset the features based on the `r ref("mlr3filters::mlr_filters_information_gain", text = "FilterInformationGain")` and then fit a `r ref("mlr3learners::mlr_learners_classif.lda", text = "LDA")`.

> We will tune the parameter `filter.frac` with a simple grid search and visualize the classification error for the different fractions.

Reviewer 2: Three times "we will" in a row, let's vary this a bit :)

> We will tune the parameter `filter.frac` with a simple grid search and visualize the classification error for the different fractions.

Reviewer 2: At this point it would be good to mention that the tuning of filter hyperpars can be fusioned with learner hyperpar tuning. This is in fact one of the big advantages of filters: they don't need an extra layer of tuning (as wrappers do).

Maybe even worth putting this into a "tip" block.

```{r optimization-060}
# reduce logging output
lgr::get_logger("bbotk")$set_threshold("warn")
lgr::get_logger("mlr3")$set_threshold("warn")

task = tsk("optdigits")

> future::plan(future::multisession(workers = 4))

Reviewer 2: Do we want to have parallel code in the book?

graph = po("filter", filter = flt("information_gain"), filter.frac = to_tune(0.1, 1)) %>>%
  po("removeconstants") %>>%
  po("scale") %>>%
  po("learner", learner = lrn("classif.lda"))

learner = as_learner(graph)

lda_tuned = auto_tuner(
  method = tnr("grid_search", resolution = 10),
  learner = learner,
  resampling = rsmp("holdout")
)
```

:::{.callout-note}

> Unless you have a very specific reason, the parameter controlling how many features are selected, should be tuned.

Reviewer 2: Maybe phrase it a bit different: "outside of tutorials, [...]" 

The average user usually doesn't know whether they have a "very specific reason" :)

:::

```{r optimization-061}
rr = resample(task, lda_tuned, rsmp("cv"), store_models = TRUE)
```

We can now extract the tuning results for each fold and visualize the dependency between the feature fraction and the classification error.

```{r optimization-062}
library("ggplot2")
archives = lapply(rr$learners,
  function(x) {
    archive = x$model$tuning_instance$archive
    archive = as.data.table(archive)
    archive[, .(information_gain.filter.frac, classif.ce)]
  }
)
archive = data.table::rbindlist(archives)

ggplot(data = archive, aes(x = as.factor(information_gain.filter.frac), y = classif.ce)) +
  geom_boxplot() +
  labs(
    title = "Optdigits task: LDA for different subset fractions",
    x = "Filter Fraction",
    y = "Classification Error"
  )
```

> We can see that using 70% - 100% of the features seems to produce fairly similar results.

Reviewer 2: There's some more beefy content required here, this does not help the user much.
Maybe 2-3 sentences describing the results and then 1-2 interpreting/discussing them (e.g. "the learner seems to produce best results with most features present which means that only few noise is in the data and most features are important) etc.

It might be worth searching for a task in which feature selection actually has an effect on the performance (and the feature space), otherwise users might think: "what do I need it for, it does not make a difference". 
This might not be easy though and if we keep it like this, we need to explain to the user what the actual point is (and why we see 70-100% here and what that means). (Note: it also highly depends on the learner).

(I know it's an example and you know that, but in the book we should address this in a somewhat scientific detail, maybe even with some references :) )
